### PR TITLE
Tree-Shake Repo

### PR DESCRIPTION
**Pull request by schmidt-sebastian**

### CODE REVIEW

File: packages/database/src/api/Query.ts [hunk 1]
```diff
@@ -48,17 +48,22 @@ import {
...
-  queryParamsStartAt,
   queryParamsEndAt,
   queryParamsEndBefore,
   queryParamsGetQueryObject,
-  queryParamsOrderBy
+  queryParamsLimitToFirst,
+  queryParamsLimitToLast,
+  queryParamsOrderBy,
+  queryParamsStartAfter,
+  queryParamsStartAt
```
Comment (Feiyang1) [inline packages/database/src/api/Query.ts:66]: What do you use to sort the imports, eslint?

File: packages/database/src/api/Reference.ts [hunk 2]
```diff
@@ -131,10 +138,11 @@ export class Reference extends Query {
     validateCallback('Reference.set', 2, onComplete, true);
 
     const deferred = new Deferred();
-    this.repo.setWithPriority(
+    repoSetWithPriority(
+      this.repo,
       this.path,
       newVal,
-      /*priority=*/ null,
+      null,
```
Comment (Feiyang1) [inline packages/database/src/api/Reference.ts:145]: nitpick - the comment still feels useful.

File: packages/database/src/core/Repo.ts [hunk 3]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+  if (repo.forceRestClient_ || beingCrawled()) {
+    repo.server_ = new ReadonlyRestClient(
+      repo.repoInfo_,
+      (
+        pathString: string,
+        data: unknown,
+        isMerge: boolean,
+        tag: number | null
+      ) => {
+        repoOnDataUpdate(this, pathString, data, isMerge, tag);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:190]:
```suggestion
        repoOnDataUpdate(repo, pathString, data, isMerge, tag);
```

File: packages/database/src/core/Repo.ts [hunk 4]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
-   * @return The time in milliseconds, taking the server offset into account if we have one.
-   */
-  serverTime(): number {
-    const offsetNode = this.infoData_.getNode(
-      new Path('.info/serverTimeOffset')
+    // Minor hack: Fire onConnect immediately, since there's no actual connection.
+    setTimeout(
+      function (connectStatus: boolean) {
+        repoOnConnectStatus(this, connectStatus);
+      }.bind(this, true),
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:199]:
```suggestion
      }.bind(repo, true),
```

File: packages/database/src/core/Repo.ts [hunk 5]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+        isMerge: boolean,
+        tag: number | null
+      ) => {
+        repoOnDataUpdate(this, pathString, data, isMerge, tag);
+      },
+      (connectStatus: boolean) => {
+        repoOnConnectStatus(this, connectStatus);
+      },
+      (updates: object) => {
+        repoOnServerInfoUpdate(this, updates);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:233]:
```suggestion
        repoOnDataUpdate(repo, pathString, data, isMerge, tag);
      },
      (connectStatus: boolean) => {
        repoOnConnectStatus(repo, connectStatus);
      },
      (updates: object) => {
        repoOnServerInfoUpdate(repo, updates);
```

File: packages/database/src/core/Repo.ts [hunk 6]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+  if (newVal === undefined) {
+    // Abort transaction.
+    transaction.unwatcher();
+    transaction.currentOutputSnapshotRaw = null;
+    transaction.currentOutputSnapshotResolved = null;
+    if (transaction.onComplete) {
+      // We just set the input snapshot, so this cast should be safe
+      const snapshot = new DataSnapshot(
+        transaction.currentInputSnapshot,
+        new Reference(this, transaction.path),
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:876]:
```suggestion
        new Reference(repo, transaction.path),
```

File: packages/database/src/core/Repo.ts [hunk 7]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+      path,
+      newNode,
+      transaction.currentWriteId,
+      transaction.applyLocally
+    );
+    eventQueueRaiseEventsForChangedPath(repo.eventQueue_, path, events);
 
-  get database(): Database {
-    return this.__database || (this.__database = new Database(this));
+    repoSendReadyTransactions(repo, this.transactionQueueTree_);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:939]:
```suggestion
    repoSendReadyTransactions(repo, repo.transactionQueueTree_);
```

File: packages/database/src/core/Repo.ts [hunk 8]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+        for (let i = 0; i < queue.length; i++) {
+          queue[i].status = TransactionStatus.COMPLETED;
+          events = events.concat(
+            repo.serverSyncTree_.ackUserWrite(queue[i].currentWriteId)
+          );
+          if (queue[i].onComplete) {
+            // We never unset the output snapshot, and given that this
+            // transaction is complete, it should be set
+            const node = queue[i].currentOutputSnapshotResolved as Node;
+            const ref = new Reference(this, queue[i].path);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:1057]:
```suggestion
            const ref = new Reference(repo, queue[i].path);
```

File: packages/database/src/core/Repo.ts [hunk 9]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
-          );
-          // There may be pending transactions that we can now send.
-          this.sendReadyTransactions_();
+        // Now remove the completed transactions.
+        repoPruneCompletedTransactionsBelowNode(
+          repo,
+          repo.transactionQueueTree_.subTree(path)
+        );
+        // There may be pending transactions that we can now send.
+        repoSendReadyTransactions(repo, this.transactionQueueTree_);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:1072]:
```suggestion
        repoSendReadyTransactions(repo, repo.transactionQueueTree_);
```

File: packages/database/src/core/Repo.ts [hunk 10]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
+      // Removing a listener can trigger pruning which can muck with
+      // mergedData/visibleData (as it prunes data). So defer the unwatcher
+      // until we're done.
+      (function (unwatcher) {
+        setTimeout(unwatcher, Math.floor(0));
+      })(queue[i].unwatcher);
+
+      if (queue[i].onComplete) {
+        if (abortReason === 'nodata') {
+          const ref = new Reference(this, queue[i].path);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:1255]:
```suggestion
          const ref = new Reference(repo, queue[i].path);
```

File: packages/database/src/core/Repo.ts [hunk 11]
```diff
@@ -269,1170 +173,1294 @@ export class Repo {
...
-   * transaction on it, or just returns the node for the given path if there are
-   * no pending transactions on any ancestor.
-   *
-   * @param path The location to start at.
-   * @return The rootmost node with a transaction.
-   */
-  private getAncestorTransactionNode_(path: Path): Tree<Transaction[]> {
-    let front;
+  // Try to send the transaction result to the server.
+  repoSendReadyTransactions(repo, this.transactionQueueTree_);
```
Comment (Feiyang1) [inline packages/database/src/core/Repo.ts:1278]:
```suggestion
  repoSendReadyTransactions(repo, repo.transactionQueueTree_);
```

**Review by Feiyang1 — CHANGES_REQUESTED** [review]

File: packages/database/src/api/Reference.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply packages/database/src/api/Reference.ts:145]: Fixed

File: packages/database/src/api/Query.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/database/src/api/Query.ts:66]: Nothing yet, this is all automated by IntelliJ via the refactor. I am planning to turn on eslint import sorting when this is all done.

**Review by schmidt-sebastian — COMMENTED** [review]
I applied your fixes, and fixed a bunch more instances of the same problem.

**Review by hsubox76 — APPROVED** [review]

**Review by Feiyang1 — APPROVED** [review]

**Pull request merged**